// sveld@0.18.1/lib/rollup-plugin.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.writeOutput = exports.generateBundle = void 0;
var fs = __importStar(require("fs"));
var fsp = __importStar(require("fs/promises"));
var path = __importStar(require("path"));
var fg = __importStar(require("fast-glob"));
var writer_ts_definitions_1 = __importDefault(require("./writer/writer-ts-definitions"));
var writer_json_1 = __importDefault(require("./writer/writer-json"));
var writer_markdown_1 = __importDefault(require("./writer/writer-markdown"));
var ComponentParser_1 = __importDefault(require("./ComponentParser"));
var get_svelte_entry_1 = require("./get-svelte-entry");
var parse_exports_1 = require("./parse-exports");
var compiler_1 = require("svelte/compiler");
var svelte_preprocess_1 = require("svelte-preprocess");
var path_1 = require("./path");
function pluginSveld(opts) {
    var result;
    var input;
    return {
        name: "plugin-sveld",
        buildStart: function () {
            input = (0, get_svelte_entry_1.getSvelteEntry)();
        },
        generateBundle: function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!(input != null)) return [3 /*break*/, 2];
                            return [4 /*yield*/, generateBundle(input, (opts === null || opts === void 0 ? void 0 : opts.glob) === true)];
                        case 1:
                            result = _a.sent();
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        },
        writeBundle: function () {
            if (input != null)
                writeOutput(result, opts || {}, input);
        }
    };
}
exports["default"] = pluginSveld;
function generateBundle(input, glob) {
    return __awaiter(this, void 0, void 0, function () {
        var dir, entry, exports, components, parser, exportEntries, _i, exportEntries_1, _a, exportName, entry_1, filePath, _b, ext, name_1, moduleName, source, processed;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    dir = fs.lstatSync(input).isFile() ? path.dirname(input) : input;
                    entry = fs.readFileSync(input, "utf-8");
                    exports = (0, parse_exports_1.parseExports)(entry, dir);
                    if (glob) {
                        fg.sync(["".concat(dir, "/**/*.svelte")]).forEach(function (file) {
                            var moduleName = path.parse(file).name.replace(/\-/g, "");
                            var source = (0, path_1.normalizeSeparators)("./" + path.relative(dir, file));
                            if (exports[moduleName]) {
                                exports[moduleName].source = source;
                            }
                        });
                    }
                    components = new Map();
                    parser = new ComponentParser_1["default"]();
                    exportEntries = Object.entries(exports);
                    _i = 0, exportEntries_1 = exportEntries;
                    _c.label = 1;
                case 1:
                    if (!(_i < exportEntries_1.length)) return [3 /*break*/, 5];
                    _a = exportEntries_1[_i], exportName = _a[0], entry_1 = _a[1];
                    filePath = entry_1.source;
                    _b = path.parse(filePath), ext = _b.ext, name_1 = _b.name;
                    moduleName = exportName;
                    if (exportEntries.length === 1 && exportName === "default") {
                        moduleName = name_1;
                    }
                    if (!(ext === ".svelte")) return [3 /*break*/, 4];
                    return [4 /*yield*/, fsp.readFile(path.resolve(dir, filePath), "utf-8")];
                case 2:
                    source = _c.sent();
                    return [4 /*yield*/, (0, compiler_1.preprocess)(source, [(0, svelte_preprocess_1.typescript)(), (0, svelte_preprocess_1.replace)([[/<style.+<\/style>/gims, ""]])], {
                            filename: path.basename(filePath)
                        })];
                case 3:
                    processed = (_c.sent()).code;
                    components.set(moduleName, __assign({ moduleName: moduleName, filePath: filePath }, parser.parseSvelteComponent(processed, {
                        moduleName: moduleName,
                        filePath: filePath
                    })));
                    _c.label = 4;
                case 4:
                    _i++;
                    return [3 /*break*/, 1];
                case 5: return [2 /*return*/, {
                        exports: exports,
                        components: components
                    }];
            }
        });
    });
}
exports.generateBundle = generateBundle;
function writeOutput(result, opts, input) {
    var inputDir = path.dirname(input);
    if ((opts === null || opts === void 0 ? void 0 : opts.types) !== false) {
        (0, writer_ts_definitions_1["default"])(result.components, __assign(__assign({ outDir: "types", preamble: "" }, opts === null || opts === void 0 ? void 0 : opts.typesOptions), { exports: result.exports, inputDir: inputDir }));
    }
    if (opts === null || opts === void 0 ? void 0 : opts.json) {
        (0, writer_json_1["default"])(result.components, __assign(__assign({ outFile: "COMPONENT_API.json" }, opts === null || opts === void 0 ? void 0 : opts.jsonOptions), { input: input, inputDir: inputDir }));
    }
    if (opts === null || opts === void 0 ? void 0 : opts.markdown) {
        (0, writer_markdown_1["default"])(result.components, __assign({ outFile: path.join(process.cwd(), "COMPONENT_INDEX.md") }, opts === null || opts === void 0 ? void 0 : opts.markdownOptions));
    }
}
exports.writeOutput = writeOutput;
